content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md

```diff
@@ -0,0 +1,545 @@
+
+<center><img src="/img/gdb4hpc-controlling-gdbs.png" width="95%" alt="Illustration showing how gdb4hpc connects to individual gdb instances. A single gdb4hpc instance is run on the login node. On each compute node, multiple application ranks are running. Each application rank has an instance of gdb attached to it. In turn, gdb4hpc remotely attaches to each individual gdb instance." title="gdb4hpc controlling multiple instances of gdb"></center>
+
+In this tutorial, you will learn how to debug a multinode MPI/CUDA application
+with gdb4hpc in the HPE Cray Programming Environment. This tutorial uses a CUDA
+application and NVIDIA GPUs as examples, but the concepts are applicable to HIP
+applications on AMD GPUs as well.
+
+## Prerequisites
+
+This tutorial assumes you already have a familiarity with command line debuggers, git, MPI, and CUDA.
+
+You will need a system with access to Cray C++ compilers and gdb4hpc, which are parts of the HPC Cray Programming Environment. You will also need access to the NVIDIA C++ compiler.
+
+## Setup
+
+To set up the debugging session, you will need to obtain the sample application
+and compile it for your chosen GPU with debug information.
+
+### Download The Sample Code
+
+For this tutorial, you will use the `simpleMPI` example from
+[NVIDIA's cuda-samples repository](https://github.com/NVIDIA/cuda-samples).
+`simpleMPI` does a simple distributed calculation involving square roots and averages:
+
+1. `simpleMPI` generates millions of random floats (2,560,000 per node)
+2. `simpleMPI` uses MPI to distribute those numbers across multiple compute nodes
+3. `simpleMPI` uses uses CUDA to calculate the square root of each number
+4. `simpleMPI` uses MPI to collect the resulting square roots and average them
+
+Start by cloning the sample code repository.
+
+```shell
+$ git clone https://github.com/NVIDIA/cuda-samples.git
+```
+
+Next, navigate to the directory containing the `simpleMPI` sample.
+
+```shell
+$ cd cuda-samples/Samples/0_Introduction/simpleMPI
+```
+
+The sample comes with some source files, some Visual Studio project files, a
+Makefile, and a README.
+
+```shell
+$ ls
+Makefile       simpleMPI.cu          simpleMPI_vs2017.vcxproj  simpleMPI_vs2022.sln
+README.md      simpleMPI.h           simpleMPI_vs2019.sln      simpleMPI_vs2022.vcxproj
+simpleMPI.cpp  simpleMPI_vs2017.sln  simpleMPI_vs2019.vcxproj
+```
+
+For this tutorial, you will only need the following source files:
+
+```shell
+simpleMPI.cpp  simpleMPI.cu  simpleMPI.h
+```
+
+Remove the other files if desired.
+
+```shell
+$ rm *.sln *.vcxproj README.md Makefile
+```
+
+### Build The Application
+
+To compile the application, load the appropriate compilers and compile the
+sample application for your chosen GPU architecture with debug information.
+
+#### Load Compilers
```

Comment:
Sentence case would have us write this as "Load compilers". If you think about it, it really changes the meaning. When you write it as "Load Compilers", "Load" appears to be an adjective to the plural noun "Compilers". When you say "Load compilers", you are giving an instruction or command.